specs/CustomDownload.md

```diff
@@ -0,0 +1,771 @@
+                                                L"Mime type: " + mimeType.get() + L"\r\n" +
+                                                L"Expected Download size in bytes: " +
+                                                std::to_wstring(expectedDownloadSizeInBytes) +
+                                                L"\r\n";
+
+                    TextInputDialog dialog(
+                        m_appWindow->GetMainWindow(), L"Download Starting", prompt.c_str(),
+                        description.c_str(), resultFilePath.get());
+                    if (dialog.confirmed)
+                    {
+                        // If user selects `OK`, the download will complete normally.
+                        // Result file path will be updated if a new one was provided.
+                        CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
+                        UpdateProgress(download.get());
+                    }
+                    else
+                    {
+                        // If user selects `Cancel`, the download will be canceled.
+                        CHECK_FAILURE(args->put_Cancel(TRUE));
+                    }
+                };
+
+                // Obtain a deferral for the event so that the CoreWebView2
+                // doesn't examine the properties we set on the event args until
+                // after we call the Complete method asynchronously later.
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
+                CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                // This function can be called to show the download dialog and
+                // complete the event at a later time, allowing the developer to
+                // perform async work before the event completes.
+                m_completeDeferredDownloadEvent = [showDialog, deferral] {
+                    showDialog();
+                    CHECK_FAILURE(deferral->Complete());
+                };
+
+                return S_OK;
+            })
+        .Get(),
+    &m_downloadStartingToken));
+}
+
+// Update download progress
+void ScenarioCustomDownloadExperience::UpdateProgress(ICoreWebView2Download* download)
+{
+    // Register a handler for the `DownloadProgressSizeInBytesChanged` event.
+    CHECK_FAILURE(download->add_DownloadProgressSizeInBytesChanged(
+        Callback<ICoreWebView2DownloadProgressSizeInBytesChangedEventHandler>(
+            [this](ICoreWebView2Download* download, IUnknown* args) -> HRESULT {
+                // Here developer can update UI to show progress of a download using
+                // `download->get_DownloadProgressSizeInBytes` and
+                // `download->get_ExpectedDownloadSizeInBytes`.
+                return S_OK;
+            })
+        .Get(),
+    &m_downloadProgressSizeInBytesChangedToken));
+
+    // Register a handler for the `DownloadStateChanged` event.
+    CHECK_FAILURE(download->add_DownloadStateChanged(
```

Comment:
We are registering for the `DownloadStateChanged` event after the download has already started. If the download was interrupted or completed before we register the event handler, we will have missed the notification entirely. In particular, since the `InterruptReason` is available only on the `DownloadStateChange` event, even if the app re-checks the State and discovers that it is `INTERRUPTED`, it can never find out why.

At least `InterruptReason` needs to be a property on the download.

If State reports the asynchronous state of the download, then you have the possibility of tearing, because the State is on the Download object, but the InterruptReason is on the event args. Suppose the download starts (InProgress) and then is quickly interrupted. When the app gets InProgress, it goes to check the State and gets Interrupted. But the InterruptReason is None since the event itself is talking about the InProgress state change. The Interrupted state change event is still waiting for its turn to come to the head of the queue.

Or does the State property not change until the StateChange event is delivered?

Later, it says that the system will sometimes auto-resume a download, which could create more tearing opportunities if the State changes asynchronously.

Reply:
We shouldn't be able to miss state changed events

Reply:
Having the state on the download class and the reason on the event args means they can be out of sync.
1. Move reason to the download class
2. Ensure the state and reason don't change until the changed event is raised (or at least that they are consistent and not the case where one has been updated and the other hasn't yet)